File: Tests/GoogleMapsUtilsTests/unit/Geometry/GMUPlacemarkTest.m
/* Copyright (c) 2016 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import <OCMock/OCMock.h>
#import <XCTest/XCTest.h>

#import "GMUPlacemark.h"

@interface GMUPlacemarkTest : XCTestCase
@end

static NSString *const kTitle = @"Placemark";
static NSString *const kSnippet = @"A test placemark.";
static NSString *const kStyleUrl = @"#test";

@implementation GMUPlacemarkTest

- (void)testInitWithGeometry {
  id geometry = OCMProtocolMock(@protocol(GMUGeometry));
  id style = OCMClassMock([GMUStyle class]);
  GMUPlacemark *placemark = [[GMUPlacemark alloc] initWithGeometry:geometry
                                                             title:kTitle
                                                           snippet:kSnippet
                                                             style:style
                                                          styleUrl:kStyleUrl];
  XCTAssertEqualObjects(placemark.geometry, geometry);
  XCTAssertEqualObjects(placemark.title, kTitle);
  XCTAssertEqualObjects(placemark.snippet, kSnippet);
  XCTAssertEqualObjects(placemark.style, style);
  XCTAssertEqualObjects(placemark.styleUrl, kStyleUrl);
}

@end

File: Tests/GoogleMapsUtilsSwiftTests/unit/Geometry/GMUPlacemarkTest.swift
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import XCTest

@testable import GoogleMapsUtils

final class GMUPlacemarkTest: XCTestCase {

    func testInitWithGeometry() {
        // 1. Create Mock Geometry (using protocol)
        let mockGeometry = MockGeometry()

        // 2. Create a GMUStyle instance
      let style = GMUStyle(styleID: "testStyle", stroke: nil, fill: nil, width: 0, scale: 1, heading: 0, anchor: .zero, iconUrl: nil, title: nil, hasFill: false, hasStroke: false)

        // 3. Create GMUPlacemark instance
        let placemark = GMUPlacemark(geometry: mockGeometry, title: "Placemark", snippet: "A test placemark.", style: style, styleUrl: "#test")

        // 4. Assert Properties
        XCTAssertIdentical(placemark.geometry, mockGeometry) // Use XCTAssertIdentical for object comparison
        XCTAssertEqual(placemark.title, "Placemark")
        XCTAssertEqual(placemark.snippet, "A test placemark.")
        XCTAssertEqual(placemark.style, style)
        XCTAssertEqual(placemark.styleUrl, "#test")
    }
}

// Mock Geometry conforming to GMUGeometry protocol
class MockGeometry: GMUGeometry {
  var type: String = ""
  
  func isEqual(_ object: Any?) -> Bool {
    guard let otherGeometry = object as? MockGeometry else {
        return false // Not the same type, so not equal
    }
    return self.type == otherGeometry.type
  }
  
  var hash: Int = 0
  
  var superclass: AnyClass?
  
  func `self`() -> Self {
    return self
  }
  
  func perform(_ aSelector: Selector!) -> Unmanaged<AnyObject>! {
    return Unmanaged.passUnretained(self)
  }
  
  func perform(_ aSelector: Selector!, with object: Any!) -> Unmanaged<AnyObject>! {
    return object as? Unmanaged<AnyObject>
  }
  
  func perform(_ aSelector: Selector!, with object1: Any!, with object2: Any!) -> Unmanaged<AnyObject>! {
    return object1 as? Unmanaged<AnyObject>
  }
  
  func isProxy() -> Bool {
    return true
  }
  
  func isKind(of aClass: AnyClass) -> Bool {
    return true
  }
  
  func isMember(of aClass: AnyClass) -> Bool {
    return true
  }
  
  func conforms(to aProtocol: Protocol) -> Bool {
    return true
  }
  
  func responds(to aSelector: Selector!) -> Bool {
    return true
  }
  
  var description: String = ""
  
    // Implement required methods of the GMUGeometry protocol (if any)
}
